---
title: "RNA-seq - LRT example"
author: "Catherine Chahrour"
date: "`r format(Sys.time(), '%d %B %Y')`"
format:
  html:
    code-fold: true
    code-tools: true
    toc: true
    toc-location: left
    embed-resources: true
    theme: cosmo 
    highlight-style: github-dark
    number-sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=F}
# Set global chunk options and load required packages quietly
knitr::opts_chunk$set(
  echo = TRUE,
  fig.align = "center",
  fig.height = 6,
  fig.width = 8,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  cache.lazy = FALSE,
  results = 'asis' 
)

# Suppress messages while loading libraries for cleaner output
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(qvalue))
suppressPackageStartupMessages(library(limma))

# Set working directory (adjust as needed)
setwd(dir = "/path/to/working_directory")
```

# Sample Information

```{r}
# Load sample metadata and convert character and numeric columns to factors
sample_info <- read.csv("full_design.csv", stringsAsFactors = TRUE) %>%
  mutate(across(where(is.character), as.factor)) %>%
  mutate(across(where(is.numeric), as.factor)) 

# Show sample metadata table
kable(sample_info, row.names = F, "simple")
```

```{r}
# Define consistent color palettes for time and treatment variables
time_col <- brewer.pal(9, "Reds")[c(5,7,9)] %>% 
  setNames(levels(sample_info$time))

treat_col <- brewer.pal(9, "Purples")[c(3,5,7,9)] %>% 
  setNames(levels(sample_info$treat))

# Define heatmap row annotation using sample metadata
row_ha <- rowAnnotation(
  time = sample_info$timepoint,
  col = list(
    time = time_col
  ),
  show_annotation_name = FALSE
)

# Define heatmap column annotation using sample metadata
col_ha <- HeatmapAnnotation(
  time = sample_info$timepoint,
  col = list(
    time = time_col,
  ),
  show_annotation_name = FALSE
)
```

# Counts from SeqNado

```{r}
# Load and clean count matrix from featureCounts output
counts <- read.delim("seqnado_output/readcounts/feature_counts/read_counts.tsv", comment = "#") %>%
  rename_with(~ gsub("(seqnado_output.aligned.)(.+)(.bam)", "\2", .), everything()) %>% 
  column_to_rownames(var = "Geneid") %>%
  dplyr::select(all_of(make.names(sample_info$sample_id))) %>%
  rename_with(~ setNames(as.character(sample_info$sample_id), make.names(sample_info$sample_id))[.]) %>%
  filter(rowSums(.) > 0) %>% 
  as.matrix()
```

# DE analysis

```{r}
# Create DESeq2 dataset using full model for timepoint
dds <- DESeqDataSetFromMatrix(
  countData = counts, 
  colData = sample_info,
  design = ~ timepoint
)

# Fit model using likelihood ratio test (LRT), removing intersect in reduced model
dds <- DESeq(dds, test="LRT", reduced = ~ 1)
```

Normalisation was performed in DESeq2 and data vst transformed to allow for visualisation.

## RLE

```{r}
# Variance stabilizing transformation followed by batch correction
vst_counts <- dds %>% 
  vst() %>% 
  assay()

# Compute RLE by subtracting row medians
rle_matrix <- vst_counts - rowMedians(vst_counts)

# Reshape and merge with sample metadata
rle_df <- rle_matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("gene") %>% 
  melt(id.vars = "gene", variable.name = "sample_id", value.name = "RLE")%>% 
  inner_join(sample_info, by = join_by(sample_id))

# Plot RLE boxplot per sample, colored by timepoint
ggplot(rle_df, aes(x = sample_id, y = RLE, fill = timepoint)) +
  geom_boxplot(outlier.size = 0.5) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = time_col) +
  labs(
    title = "Relative Log Expression (RLE) Plot", 
    y = "RLE (log2 scale)", 
    x = "Sample"
  )
```

## Sample distance matrix

```{r}
# Compute sample-to-sample Euclidean distance matrix
sampleDists <- vst_counts %>% 
  t() %>% 
  dist() %>% 
  as.matrix()

# Plot sample distance matrix as heatmap
set.seed(123)
HM_sampleDists <- Heatmap(
  sampleDists, 
  top_annotation = col_ha,
  left_annotation = row_ha,
  row_title = NULL,
  show_row_names = TRUE,
  cluster_rows = TRUE,
  column_title = NULL, 
  show_column_names = FALSE,
  cluster_columns = TRUE,
  show_heatmap_legend = FALSE,
  col = colorRamp2(c(0, max(sampleDists)), c("black", "white"))
)

draw(HM_sampleDists, merge_legend = TRUE)
```

## Overview by PCA {.tabset}

```{r}
# Perform PCA and calculate percent variance per PC
pca_res <- vst_counts %>% 
  t() %>% 
  prcomp()

percentVar <- round(100 * pca_res$sdev^2 / sum(pca_res$sdev^2), digits = 2)

# Merge with metadata
pca_data <- pca_res$x %>% 
  as.data.frame() %>% 
  rownames_to_column("sample_id") %>%
  inner_join(sample_info, by = join_by(sample_id))

# PCA plot with sample labels
ggplot(pca_data, aes(PC1, PC2, color = timepoint)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  geom_label_repel(
    aes(label = replicate), 
    color = "black", 
    fill = "white",
    size = 2,
    max.overlaps = 20,
    show.legend = FALSE) +
  scale_color_manual(values = time_col) +
  ggtitle("PCA - PC1 vs PC2") +
  theme_light() 
```

## PCA correlate with covariates

```{r}
# Convert metadata to numeric factors for correlation with PCs
metadata_for_cor <- colData(dds) %>%
  as.data.frame() %>% 
  dplyr::select(-c(sample_name)) %>% 
  mutate_if(is.character, as.factor) %>% 
  mutate_if(is.factor, as.integer)

# Correlation between metadata and first 8 PCs using Spearman method
pca_meta_corr <- abs(cor(
  metadata_for_cor, 
  pca_data[, paste0("PC", 1:8)], 
  method = "spearman"))

# Plot correlation heatmap
heatmap <- Heatmap(
  pca_meta_corr, 
  name = "Correlation", 
  col = colorRamp2(
    seq(0, 1, length = 9), 
    brewer.pal(n = 9, name = "Blues")), 
  cluster_rows = FALSE,
  cluster_columns = FALSE, 
  width = ncol(pca_meta_corr)*unit(1, "cm"), 
  height = nrow(pca_meta_corr)*unit(1, "cm"),
  heatmap_legend_param = list(
    title = "Correlation",
    title_position = "topcenter",
    direction = "horizontal")
)

draw(heatmap, heatmap_legend_side = "bottom", annotation_legend_side = "right")
```


# Pairwise comparisons {.tabset}

```{r}
colors <- brewer.pal(5, "RdBu")[c(5,1)] %>% 
  append("grey") %>% 
  set_names(c("Down", "Up", "NotSig"))

DE_summary_fn <- function(df, fcs, pvals) {
  summary_df <- data.frame()
  for (fc in fcs) {
    for (pval in pvals) {
      up_count = sum(df$padj < pval & df$log2FoldChange >  fc)
      down_count = sum(df$padj < pval & df$log2FoldChange < -fc)

      summary_df[1, paste0('log2FC>', fc, '_', pval)] <- up_count
      summary_df[1, paste0('log2FC<', fc, '_', pval)] <- down_count
    }
  }

  summary_df %>%
    pivot_longer(cols = starts_with("log"), values_to = "count") %>%
    separate(col = name, into = c("log2FoldChange", "pvalue"), sep = "_") %>%
    mutate(
      FC_direction = ifelse(str_detect(log2FoldChange, ">"), "up", "down"), 
      FoldChange_threshold = str_extract(log2FoldChange, "(?<=<|>)(\\d+)"), 
      count = ifelse(str_detect(log2FoldChange, ">"), count, -count)
    )
}

fold_changes <- c(0:3)
pvalues <- c(0.05, 0.01, 0.001)
res_names <- resultsNames(dds)[-1]

DE_results_pairwise <- list()

for(comp in res_names) {
  cat("## ", comp, "\n \n")
  
  res <- results(dds, name = comp, test="Wald") %>% 
   as.data.frame() %>%
    rownames_to_column("gene") %>% 
    arrange(padj) %>% 
    filter(!is.na(padj))
  
  pval <- ggplot(res, aes(pvalue)) +
    geom_histogram(fill = "lightblue", binwidth = 0.01) +
    theme_light() +
    theme(text = element_text(size = 15)) +
    ggtitle(paste("P-value distribution: ", comp)) 
  print(pval)
  
  res_filtered <- res %>% 
    filter(padj < 0.05)
  
  res_list <- list(res_filtered)
  names(res_list) <- comp
  DE_results_pairwise <- append(DE_results_pairwise, res_list)
  DEsummary <- DE_summary_fn(res, fold_changes, pvalues)

  p <- ggplot(DEsummary, aes(x=pvalue, y=count, 
                        fill=FoldChange_threshold,
                        group=FoldChange_threshold)) +
    geom_bar(stat="identity", position=position_dodge()) +
    geom_hline(yintercept=0, color = "grey") +
    ggtitle(paste("DE genes:", comp)) +
    scale_fill_brewer(palette = "Dark2") +
    theme_light() +
    theme(text = element_text(size = 15))
  print(p)

  res4plot <- res %>%
    dplyr::select(gene, log2FoldChange, padj) %>%
    arrange(padj) %>%
    mutate(
      DE = case_when(
        padj < 0.05 & log2FoldChange < 0 ~ "Down",
        padj < 0.05 & log2FoldChange >= 0 ~ "Up",
        TRUE ~ "NotSig"
      ),
      label = if_else(gene %in% union(
        head(filter(., DE == "Up")$gene, 5),
        head(filter(., DE == "Down")$gene, 5)
      ), gene, as.character(NA))
    )

  maxLFC <- max(abs(res4plot$log2FoldChange))
  maxpadj <- max(-log10(res4plot$padj), na.rm = TRUE)
  
  
  g <- ggplot(data=res4plot, aes(x=log2FoldChange, y=-log10(padj), col=DE, label=label)) +
    geom_point(size=1) +
    theme_minimal() +
    geom_label_repel(aes(label = label), show.legend = F, max.overlaps = 20) +
    scale_colour_manual(values = colors) +
    geom_vline(xintercept=0, linetype='dotted', col="black") +
    geom_hline(yintercept=-log10(0.05), linetype='dotted', col="black") +
    xlab("log2 fold change") +
    ylab("-log10 adjusted p-value") +
    xlim(-maxLFC, maxLFC) +
    ylim(0, maxpadj) +
    ggtitle(comp) +
    theme(text = element_text(size = 15))
  outname <- paste0("plots/volcano_", comp,".pdf")
  ggsave( outname, g, width = 8, height = 6)
  print(g)
  
  cat(" \n \n")
}
```

# Overlapping genes in comparisons

```{r}
library(venn)
venn::venn(
  x = list(
    "01_hour" = DE_results_pairwise$timepoint_01_vs_00$gene,
    "02_hour" = DE_results_pairwise$timepoint_02_vs_00$gene
  ),
  zcolor = time_col[c("het", "hom")],
  opacity = 0.5,
  ilcs = 1,
  ilabels = "counts",
  box = F,
)
```


# Session Info

```{r results='markup'}
sessionInfo()
```

